### PR TITLE
Add .clang-format file

### PR DESCRIPTION
**Pull request by vtjnash**
Based on #15591, but further updated to be closer to our current style. I don't plan to merge the format diff (just the first commit with the format description), but will do that separately as files are being touched. It's included here for reference on how these rules would apply to stylize the code.

I took out the makefile changes, since `clang-format -i <filename>` seemed easy enough to use on it's own. You can also configure your favorite vim / bbedit / VS / emacs / ed(itor) to apply formatting on save or on keystroke: https://clang.llvm.org/docs/ClangFormat.html

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: true
+BinPackArguments: false
+BinPackParameters: false
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Stroustrup # Attach
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+ColumnLimit:     80
```
Comment (tkelman) [inline .clang-format]: 92

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+CommentPragmas:  '^ IWYU pragma:'
+ConstructorInitializerAllOnOneLineOrOnePerLine: false
+ConstructorInitializerIndentWidth: 4
+ContinuationIndentWidth: 8
```
Comment (tkelman) [inline .clang-format]: 8 is too much except for continuations of block openers like function signatures or if conditionals, and this didn't apply to all of those

File: src/APInt-C.h [hunk 3]
```diff
@@ -9,81 +9,121 @@ extern "C" {
...
-JL_DLLEXPORT int LLVMSub_uov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMSub_sov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMMul_sov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMMul_uov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMDiv_sov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMDiv_uov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMRem_sov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
-JL_DLLEXPORT int LLVMRem_uov(unsigned numbits, integerPart *pa, integerPart *pb, integerPart *pr);
+JL_DLLEXPORT int
+LLVMICmpSLT(unsigned numbits, integerPart *pa, integerPart *pb);
```
Comment (ararslan) [inline src/APInt-C.h]: Was this auto-changed or something? IMO this looks really weird and bad.

File: .clang-format (hunk 2 above)
Comment (vtjnash) [inline .clang-format]: Split "if" conditionals will usually indented by `AlignOperands` and ignore this. Blocks are indented by `IndentWidth`, not this. Function signatures and parameters (if wrapped) should have this applied, as should any statement that wraps to the next line (the other common case being after an equals sign).

I like setting this fairly high, as an encouragement to be splitting those arguments into variables with real names.

Quote the Kernel:

> Now, some people will claim that having 8-character indentations makes the code move too far to the right, and makes it hard to read on a 80-character terminal screen. The answer to that is that if you need more than 3 levels of indentation, you’re screwed anyway, and should fix your program.
(https://www.kernel.org/doc/html/v4.10/process/coding-style.html)

File: src/APInt-C.h (hunk 3 above)
Comment (vtjnash) [inline src/APInt-C.h]: yes. I don't plan to commit these auto-format results, so that some hand-customization may be done

Comment (ararslan) [inline src/APInt-C.h]: Okay, thanks for explaining

File: .clang-format (hunk 2 above)
Comment (vtjnash) [inline .clang-format]: Actually, I checked and continuations are being handled by AlignAfterOpenBracket, so it's unfortunately coupled between whether blocks are aligned (AlignAfterOpenBracket: DontAlign) and whether the first parameter is aligned with the other parameters (AlignAfterOpenBracket: AlwaysBreak). It seems like whoever made that a ternary (circa 3.5) didn't notice this, or perhaps wanted it this way.

**Comment by Keno** [comment]
There's also the `git-clang-format` utility, which is very helpful to just format the things that you changed.

File: .clang-format (hunk 2 above)
Comment (tkelman) [inline .clang-format]: there are definitely if conditions that are indented to the same level as the following control flow block in this diff, so AlignOperands either isn't doing that or isn't accomplishing the goal of distinguishing the line continuation from the control flow

if we wanted 8-space indents like the kernel, we'd use that uniformly for all indentation. as it is, it's a mix, mostly 4 spaces, and the parts that use 4 spaces look much better when the next complete line is within the same block.

Comment (vtjnash) [inline .clang-format]: Misreading, I realize that was misleading. The kernel uses 8-column tabs for block indenting. For continuations, the only guidance they have seemed to be "substantially" (which I assumed to mean more than 8, but this also wasn't relevant for my point)

**Comment by StefanKarpinski** [comment]
If we're going to do this, why not just apply the new standard formatting everywhere all at once?

**Comment by maleadt** [comment]
https://github.com/JuliaLang/julia/pull/35784

**Pull request closed without merging**